File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/ThrownWeaponTraitEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum ThrownWeaponTraitEnum implements WeaponTraitEnum {
    THROWN_10FT(10),
    THROWN_20FT(20);

    private final Integer value;
    private final static Map<Integer, ThrownWeaponTraitEnum> CONSTANTS = new HashMap<>();

    static {
        for (ThrownWeaponTraitEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    ThrownWeaponTraitEnum(Integer value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return String.format("Thrown %d feet", this.value);
    }

    @JsonValue
    public Integer value() {
        return this.value;
    }

    @JsonCreator
    public static ThrownWeaponTraitEnum fromValue(Integer value) {
        ThrownWeaponTraitEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value.toString());
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/ThrownWeaponTraitEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class ThrownWeaponTraitEnum(private val value: Int) : WeaponTraitEnum {
    THROWN_10FT(10), THROWN_20FT(20);

    companion object {
        private val CONSTANTS: MutableMap<Int, ThrownWeaponTraitEnum> = HashMap()

        @JsonCreator
        fun fromValue(value: Int): ThrownWeaponTraitEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value.toString())
        }

        init {
            for (c in values()) {
                CONSTANTS[c.value] = c
            }
        }
    }

    override fun toString(): String {
        return String.format("Thrown %d feet", value)
    }

    @JsonValue
    fun value(): Int {
        return value
    }

}